elastic/kibana
# [Fleet] Allow to preconfigure alternative ES outputs (on the same cluster)

## Summary

Resolve partially #108968

Allow to preconfigure alternative Elasticsearch output  in the same cluster and use them in the preconfigure API (= specifying policy and output in the kibana config file).

That PR allow to define multiple outputs in the Kibana config file and to use them in preconfigured policies as the data output or monitoring output.

Due to a limitation in Fleet server we only support one output per agent policy, this is enforced in the config schema.

We will not support remote cluster in 7.16.0 so that PR remove the ability to specify `fleet_server.service_token`.


Predefined ouput can only be added/updated/removed through the kibana config file.

## Implementation details

* I move the generation of the full policy (aka aggregated policy for the agent in his own module, as the agent policy started to be really large) 

### Config examples 

```yaml
xpack.fleet.outputs:
  - name: 'Test output'
    type: 'elasticsearch'
    id: 'output-123'
    hosts: ['http://test.fr']
  - name: 'Test output 2'
    type: 'elasticsearch'
    id: 'output-test'
    hosts: ['http://test.fr']
xpack.fleet.agentPolicies:
  - name: Alternative Default Agent Policy
    id: policy124
    namespace: test
    is_default: true
    is_managed: true
    data_output_id: output-123
    monitoring_output_id: output-123
    package_policies:
      - package:
          name: system
        name: System Integration
  - name: Alternative Default Fleet server Agent Policy
    id: policy123
    namespace: test
    is_default_fleet_server: true
    package_policies:
      - package:
          name: system
        name: System Integration
  - name: Test agent policy
    is_managed: true
    id: test-policy
    is_default_fleet_server: true
    monitoring_enabled: []
    package_policies:
      - package:
          name: fleet_server
        name: Fleet Server
        inputs:
          - type: fleet-server
            keep_enabled: true
```            

### How to test

You can build your own fleet server based on https://github.com/elastic/fleet-server/pull/713 or just check for the generated policy in `.fleet-policies`


I am planning to add an e2e test after this PR and the fleet server PR got merged.
